src/main/java/hexgui/gui/HexGui.java

```diff
@@ -0,0 +1,2393 @@
+  private boolean askSaveGame() {
+    Object options[] = {"Save", "Discard", "Cancel"};
+    int n =
+        JOptionPane.showOptionDialog(
+            this,
+            "Game has changed.  Save changes?",
+            "Save Game?",
+            JOptionPane.YES_NO_CANCEL_OPTION,
+            JOptionPane.QUESTION_MESSAGE,
+            null,
+            options,
+            options[0]);
+    if (n == 0) {
+      if (cmdSaveGame()) return true;
+      return false;
+    } else if (n == 1) {
+      return true;
+    }
+    return false;
+  }
+
+  /** Saves the current game state as a position in the specified sgf file. */
+  private boolean savePosition(File file) {
+    Node root = new Node();
+
+    GameInfo info = new GameInfo();
+    info.setBoardSize(m_guiboard.getBoardSize());
+    m_guiboard.storePosition(root);
+    root.setPlayerToMove(m_tomove);
+    return save_tree(file, root, info);
+  }
+
+  /**
+   * Save game to file.
+   *
+   * @return true If successful.
+   */
+  private boolean save(File file) {
+    return save_tree(file, m_root, m_gameinfo);
+  }
+
+  private boolean save_tree(File file, Node root, GameInfo gameinfo) {
+    FileOutputStream out;
+    try {
+      out = new FileOutputStream(file);
+    } catch (FileNotFoundException e) {
+      ShowError.msg(this, "File not found!");
+      return false;
+    }
+
+    new SgfWriter(out, root, gameinfo);
+    return true;
+  }
+
+  /* Load game from file. */
+  private SgfReader load(File file) {
+    FileInputStream in;
+    try {
+      in = new FileInputStream(file);
+    } catch (FileNotFoundException e) {
+      ShowError.msg(this, "File not found!");
+      return null;
+    }
+
+    SgfReader sgf;
+    try {
+      sgf = new SgfReader(in);
+    } catch (SgfReader.SgfError e) {
+      ShowError.msg(this, "Error reading SGF file:\n \"" + e.getMessage() + "\"");
+      return null;
+    }
+    return sgf;
+  }
+
+  // ------------------------------------------------------------
+
+  /**
+   * Show save dialog, return File of selected filename.
+   *
+   * @return null If aborted.
+   */
+  private File showSaveAsDialog() {
+    JFileChooser fc = new JFileChooser(m_preferences.get("path-save-game"));
+    if (m_file != null) fc.setSelectedFile(m_file);
+    int ret = fc.showSaveDialog(this);
+    if (ret == JFileChooser.APPROVE_OPTION) return fc.getSelectedFile();
+    return null;
+  }
+
+  /**
+   * Show open dialog, return File of selected filename.
+   *
+   * @return null If aborted.
+   */
+  private File showOpenDialog() {
+    JFileChooser fc = new JFileChooser(m_preferences.get("path-load-game"));
+    int ret = fc.showOpenDialog(this);
+    if (ret == JFileChooser.APPROVE_OPTION) return fc.getSelectedFile();
+    return null;
+  }
+
+  // ------------------------------------------------------------
+
+  private void acquireSemaphore() {
+    try {
+      m_semaphore.acquire();
+    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)